### PR TITLE
AGDIGGER-58 - Digger java client backbone

### PR DESCRIPTION
**Pull request by wtrocki**
Backbone for java client api.

Supports:
- maven builds
- official jenkins java library included
- sample project create implementation

### CODE REVIEW

File: src/main/java/com/redhat/digkins/services/CreateJobService.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package com.redhat.digkins.services;
```
Comment (aliok) [inline src/main/java/com/redhat/digkins/services/CreateJobService.java:1]: @wtrocki we should use org.aerogear namespace I think. Or, org.jboss.aerogear

An example: <https://github.com/aerogear/aerogear-android-authz/tree/master/aerogear-android-authz/src/main/java/org/jboss/aerogear/android/authorization/oauth2>

@matzew what do you say?

File: src/main/java/com/redhat/digkins/DiggerClient.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * Digger Java Client
+ * <p>
+ * Interact with digger jenkins api!
+ */
+public class DiggerClient {
+
+  private final JenkinsServer jenkins;
+
+  public DiggerClient(JenkinsAuth auth) throws URISyntaxException {
+    this.jenkins = new JenkinsServer(new URI(auth.getUrl()), auth.getUser(), auth.getPassword());
```
Comment (aliok) [inline src/main/java/com/redhat/digkins/DiggerClient.java:21]: +1 on creating the `java.net.URI` here instead of later. I was gonna suggest that yesterday :)

File: src/main/java/com/redhat/digkins/DiggerClient.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Create new digger job on jenkins platform
+   *
+   * @param name      - job name that can be used later to reference job
+   * @param gitRepo   - git repository url (full git repository url. e.g git@github.com:wtrocki/helloworld-android-gradle.git
+   * @param gitBranch - git repository branch (default branch used to checkout source code)
+   */
+  public void createJob(String name, String gitRepo, String gitBranch) throws DiggerClientException {
+    CreateJobService service = new CreateJobService(this.jenkins);
```
Comment (aliok) [inline src/main/java/com/redhat/digkins/DiggerClient.java:32]: To me, a service is stateless and it is safe to have it as a field on the class.
`CreateJobService` is stateless indeed: state is in the dependencies.
Why creating it every time then? Just keep it as a field on DiggerClient.

File: src/main/java/com/redhat/digkins/util/DiggerClientException.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  public DiggerClientException(String message, Throwable cause) {
+    super(message, cause);
+  }
+
+  public DiggerClientException(Throwable cause) {
+    super(cause);
+  }
+
+  @Override
+  public Throwable fillInStackTrace() {
```
Comment (aliok) [inline src/main/java/com/redhat/digkins/util/DiggerClientException.java:27]: @wtrocki what is the reason of overriding this method?
I've never seen it before :)

File: src/main/java/com/redhat/digkins/DiggerClient.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Create client using provided url and credentials
+   *
+   * @param url      - jenkins url
+   * @param user     - jenkins user
+   * @param password - jenkins password
+   * @return client instance
+   */
+  public static DiggerClient from(String url, String user, String password) throws DiggerClientException {
```
Comment (aliok) [inline src/main/java/com/redhat/digkins/DiggerClient.java:48]: What is the purpose of having this static builder? Normally, you do it when in one of the cases:
* too many parameters in constructor
* want to hide the constructor because you don't want others to extend a class
* want to hide the constructor because you want a singleton

If we want any of these, I'd say let's make the constructor private (or package-private or protected). If not, let's delete the static builder.

File: src/main/java/com/redhat/digkins/DiggerClient.java (hunk 3 above)
Comment (wtrocki) [reply src/main/java/com/redhat/digkins/DiggerClient.java:32]: Good point of view. I'm using here short lived objects because of the way we going to use this library. When creating client we would use it to do only one job (just calling one Service).Having all services created upfront as part of DiggerClient would mean that we would require x objects (where x is number of services) where actually only one Service would be required. Happy to change this approach - no strong feelings here.

File: src/main/java/com/redhat/digkins/util/DiggerClientException.java (hunk 4 above)
Comment (wtrocki) [reply src/main/java/com/redhat/digkins/util/DiggerClientException.java:27]: What we doing here is called lightweight exception and is typical pattern for applications who choose to use exceptions to direct application flow. Normally java apps return exceptions if something bad is happening and it would require some stack trace check later. In some cases developers may chose exceptions as way to define application flow. For example: `CustomerAlreadyRegisteredException`. In our case library we using is using that pattern so I'm wrapping their exceptions into our for better error handling. That exception may appear during the normal application execution. Statement above is used to not pollute logs with stack trace because it has absolutely no actual meaning here. I have no strong feeling to have it.

TL;DR; That's typical pattern for Client based apps when you using Exceptions as way to communicate with users code.

File: src/main/java/com/redhat/digkins/DiggerClient.java (hunk 5 above)
Comment (wtrocki) [reply src/main/java/com/redhat/digkins/DiggerClient.java:48]: Having this private would be ok.
